Add physics helper (#4)
turpaan64/z64hdr: helpers/README.md

# Z64 Helpers

By default these aren't set `static inline`. To do this, you can `#define` `_Z64HDR_HELPER_PREFIX_` to achieve this.
```c
#define _Z64HDR_HELPER_PREFIX_ static inline
#include <helpers/Helper.h>
```
---
### List of content
1. [MATRIX](#matrix)
	1. [Matrix Multiply](#matrix-mult)
	1. [Matrix Macros](#matrix-macros)
1. [PHYSICS](#physics)
	1. [PhysicsStrand Struct](#physicsstrand-struct)
		1. [PhysicsInfo](#physicsinfo)
		1. [PhysicsHead](#physicshead)
		1. [PhysicsGfx](#physicsgfx)
		1. [PhysicsSpheres](#physicsspheres)
		1. [PhysicsConstraint](#physicsconstraint)
		1. [PhysicsRigidity](#physicsrigidity)
    1. [Examples](#examples)
        1. [PhysicsStrand](#physicsstrand)
        1. [ActorInit](#actorinit)
        1. [OverridePostDraw](#overridepostdraw)
		1. [PhysicCallback](#physiccallback)
        1. [ActorDraw](#actordraw)
1. [Credits](#credits)
---
# Matrix
## Matrix Mult
```c

> void Matrix_MultX(f32 scale, Vec3f* dst);

Reviewer: ```suggestion
void Matrix_MultX(f32 scale, Vec3f* dst); // dst = curMatrix * (scale, 0, 0, 1)
```
you may want to change `scale` in these to x, y, z

void Matrix_MultY(f32 scale, Vec3f* dst);
void Matrix_MultZ(f32 scale, Vec3f* dst);
```
[Back to top](#list-of-content)

---
## Matrix Macros
```c
#define Matrix_RotateY_s(binang, A) Matrix_RotateY(BINANG_TO_RAD(binang), A)
#define Matrix_RotateX_s(binang, A) Matrix_RotateX(BINANG_TO_RAD(binang), A)
#define Matrix_RotateZ_s(binang, A) Matrix_RotateZ(BINANG_TO_RAD(binang), A)

#define Matrix_RotateY_f(degf, A)   Matrix_RotateY(DEGF_TO_RADF(degf), A)
#define Matrix_RotateX_f(degf, A)   Matrix_RotateX(DEGF_TO_RADF(degf), A)
#define Matrix_RotateZ_f(degf, A)   Matrix_RotateZ(DEGF_TO_RADF(degf), A)
```
[Back to top](#list-of-content)

---
# Physics
---
## PhysicsStrand Struct
```md
typedef struct {
	PhysicsInfo       info;
	PhysicsHead       head;
	PhysicsGfx        gfx;
	PhysicsSpheres    spheres; // "collision" spheres
	PhysicsConstraint constraint;
	PhysicsRigidity   rigidity;
	f32 limbsLength[];

> } PhysicsStrandInit;

Reviewer: would be nice to add information, but if it's just comments may as well leave them in the source rather than having to maintain duplicates of the structure definition (same for all structs in the readme)

---
### PhysicsHead
```c
typedef struct {
	Vec3f pos;
	Vec3s rot;
	MtxF* mtxF;
} PhysicsHead;
```
[Back to top](#list-of-content)

---
### PhysicsGfx
```c
typedef struct {
	u32   dlist;
	Vec3f scale;  // Gfx scale
	u8    segID;  // For matrix
	u8    noDraw; /* If there needs to be calculations
	                 to get in position before drawing */
} PhysicsGfx;
```
[Back to top](#list-of-content)

---
### PhysicsSpheres
```c
typedef struct {
	s32    num; // amount of spheres
	Vec3f* centers;
	f32    radius;
} PhysicsSpheres;
```
[Back to top](#list-of-content)

---
### PhysicsConstraint
```c
typedef struct {
	u8    lockRoot;    // Prevent physics rotating root limb
	Vec2f rotStepCalc; // DEG, limits rot to next limb in main calc
	Vec2f rotStepDraw; // DEG, limits rot on draw, smoothens output
} PhysicsConstraint;
```
[Back to top](#list-of-content)

---
### PhysicsRigidity
```c
typedef struct {
	s32   num;  // How many limbs will be smoothed with push
	Vec3f push; // direction Z, pushes based on rot[0]
	f32   mult; // How much the pushing fill affect
	Vec3f rot;  // DEG, rigids towards, relative rot
} PhysicsRigidity;
```
[Back to top](#list-of-content)

---
## Examples

> Here you set the arrays for physics to store/read values from. In this specific example the substruct is named `PonytailPhysics`, but you can name it whatever you like.

Reviewer: ```suggestion
Here you set the arrays for physics to store/read values from. In this specific example the substruct is named `PonytailPhysics`, but you can name it whatever you like.
```
`PonytailPhysics` ?

Author: There might be some oversights in readme. I'll do another check at it today!

	Vec3f         bodyPartsPos[6];
	Vec3f         tailEnd;
} EnNPC;
```
[Back to top](#list-of-content)

---
### PhysicsStrand
Here are all the options that affect how the strand will be processed. These could also be stored in Actors struct, if these are set into it in init.
```c
PhysicsStrandInit sPonytailInit = {
	.info = {
		.numLimbs = 9,
		.gravity = -2.0f,
		.floorY = 2.0f,
		.maxVel = 64.0f,
		.velStep = 0.7f,
		.velMult = 0.9f,
	},
	.gfx = {
		.dlist = DL_PONYTAIL,
		.scale = {
			0.0055f,
			0.0055f,
			0.0055f
		},
		.segID = 0x0B,
		.noDraw = true,
	},
	.spheres = {
		.num = 6,
		.centers = NULL,
		.radius = 4.0f
	},
	.constraint = {
		.lockRoot = false,
		.rotStepCalc = {
			.x = 45.0f,
			.y = 45.0f,
		},
		.rotStepDraw = {
			.y = 25.0f,
			.x = 0.0f,
		},
	},
	.rigidity = {
		.num = 5,
		.push = {
			0,
			0.2f,
			-3.0f
		},
		.mult = 0.2f,
		.rot = {
			.x = 50.0f,
			.y = 0.0f,
			.z = 0.0f,
		},
	},
	.limbsLength = {
		-930.0f,
		-1530.0f +  930.0f,
		-2160.0f + 1530.0f,
		-3150.0f + 2160.0f,
		-4090.0f + 3150.0f,
		-5190.0f + 4090.0f,
		-6070.0f + 5190.0f,
		-6810.0f + 6070.0f,
		-400.0f,
	},
};
```
[Back to top](#list-of-content)

---

> ### ActorInit

Reviewer: ```suggestion
### Actor Init function
```
to avoid confusion with ActorInit, the struct for initvars (may want to edit the other titles too)

```c
void EnNpc_Init(EnNpc* this, GlobalContext* globalCtx) {
	Actor_SetScale(&this->actor, 0.0055f);
	SkelAnime_InitFlex(globalCtx, &this->skelAnime, SKEL_NPC, ANIM_IDLE, this->jointTable, this->morphTable, SKEL_NPC_NUMBONES_DT);
	
	Physics_SetPhysicsStrand(&sPonytailInit, &this->phyPonytail, this->phyJointLength, this->bodyPartsPos);
}
```
[Back to top](#list-of-content)

---
### OverridePostDraw

> In `SkelAnime` PostDraw you can get the "sphere" collision centers like so. This will prevent the limbs of the strand go through them.

Reviewer: ```suggestion
In `SkelAnime` PostDraw you can update the "sphere" collision centers like so (`bodyPartsPos`). This will prevent the limbs of the strand from going through the skeleton limbs.
```

```c
void EnNpc_PostDraw(GlobalContext* globalCtx, s32 limbIndex, Gfx** dList, Vec3s* rot, void* thisx) {
	EnNpc* this = (EnNpc*)thisx;
	Vec3f zero = { 0 };
	
	switch (limbIndex) {
		case NpcLimb_Head:
			Matrix_MultVec3f(&zero, &this->actor.focus.pos);
			break;
		case NpcLimb_Ponytail1:
			Physics_GetHeadProperties(&this->phyPonytail, &zero, 0);
			break;
	    case NpcLimb_Upperbody:

> 		    Matrix_MultVec3f(&zero, &this->bodyPartsPos[0]);

Reviewer: indentation a bit broken here

		    break;
	    case NpcLimb_Upperarm_R:
		    Matrix_MultVec3f(&zero, &this->bodyPartsPos[1]);
		    break;
	    case NpcLimb_Thigh_R:
		    Matrix_MultVec3f(&zero, &this->bodyPartsPos[2]);
		    break;
	    case NpcLimb_Waist:
		    Matrix_MultVec3f(&zero, &this->bodyPartsPos[3]);
		    break;
	    case NpcLimb_Upperarm_L:
		    Matrix_MultVec3f(&zero, &this->bodyPartsPos[4]);
		    break;
	    case NpcLimb_Thigh_L:
		    Matrix_MultVec3f(&zero, &this->bodyPartsPos[5]);
		    break;
	}
}
```
[Back to top](#list-of-content)

---
### PhysicCallback
This callback can have up to two `pointers` of your choice. Type of the pointers are declared in this callback function itself and then assigned to the `Physics_DrawDynamicStrand` function if used.
```c
void EnNpc_PhysicsCallback(s32 limbIndex, GlobalContext* globalCtx, void* arg2) {
	Vec3f zero = { 0.0f, 100.0f, 0.0f };

	if (limbIndex == LIMB_END)
		Matrix_MultVec3f(&zero, &this->tailEnd);
}
```
[Back to top](#list-of-content)

---
### ActorDraw

> In draw you can do a `for loop` to get the positions calculated instantly. Also remember to use the extra arguments if you're going to use `PhysicCallback`.

Reviewer: wouldn't it make more sense to move that loop to the init function

Author: Yeah it would, but it requires the head properties in order to get correct results. One way could be taking SkelAnime_DrawFlexOpa and using it in Init maybe.

Reviewer: didn't think of that, yeah I guess moving to init would be more trouble than what it's worth. It's fine as is

```c
void EnNPC_Draw(EnNPC* this, GlobalContext* globalCtx) {
	SkelAnime_DrawFlexOpa(
		globalCtx,
		this->skelAnime.skeleton,
		this->skelAnime.jointTable,
		this->skelAnime.dListCount,
		NULL,
		EnNPC_PostDraw,
		&this->actor
	);
	
	if (this->phyPonytail.gfx.noDraw == true) {
		for (s32 i = 0; i < 60; i++) {
			Physics_DrawDynamicStrand(
				globalCtx->state.gfxCtx,
				&POLY_OPA,
				this->phyJointTable,
				&this->phyPonytail,
				0,
				0,
				0
			);
		}
		this->phyPonytail.gfx.noDraw = false;
	}
	
	Physics_DrawDynamicStrand(
		globalCtx->state.gfxCtx,
		&POLY_OPA,
		this->phyJointTable,
		&this->phyPonytail,
		EnNpc_PhysicsCallback,
		globalCtx,

> 		0

Reviewer: ```suggestion
		NULL
```

# Z64 Helpers

By default these aren't set `static inline`. To do this, you can `#define` `Z64HDR_HELPER_QUALIFIERS` to achieve this.
```c
#define Z64HDR_HELPER_QUALIFIERS static inline
#include <helpers/HelperOfYourChoice.h>
```
### Table of Content
 - [1. MATRIX](#matrix)
	 - [1.1 Matrix Multiply](#matrix-mult)
	 - [1.2 Matrix Macros](#matrix-macros)
 - [2. PHYSICS](#physics)
	- [2.1 PhysicsStrand Struct](#physicsstrand-struct)
		- [2.1.1 PhysicsInfo](#physicsinfo)
		- [2.1.2 PhysicsHead](#physicshead)
		- [2.1.3 PhysicsGfx](#physicsgfx)
		- [2.1.4 PhysicsSpheres](#physicsspheres)
		- [2.1.5 PhysicsConstraint](#physicsconstraint)
		- [2.1.6 PhysicsRigidity](#physicsrigidity)
	- [2.2 Examples](#examples)
		- [2.2.1 PhysicsStrandInit](#physicsstrandinit)
		- [2.2.2 Actor Init function](#actor-init-function)
		- [2.2.3 OverridePostDraw](#overridepostdraw)
		- [2.2.4 PhysicCallback](#physiccallback)
		- [2.2.5 Actor Draw function](#actor-draw-function)
- [3. Credits](#credits)
# Matrix
## Matrix Mult

> These are pretty much same as `Matrix_MultVec3f`, but instead of using Vec3f, these are for individual axises.

Reviewer: ```suggestion
These are pretty much the same as `Matrix_MultVec3f`, but instead of using Vec3f, these are for individual axes.
```

```c
void Matrix_MultX(f32 x, Vec3f* dst);
void Matrix_MultY(f32 y, Vec3f* dst);
void Matrix_MultZ(f32 z, Vec3f* dst);
```
[Back to Table of Content](#table-of-content)

## Matrix Macros

> These macros are for using different format of rotations. By default `Matrix_Rotate` first argument is used with radiance.

Reviewer: ```suggestion
These macros are for using different units of rotations. By default `Matrix_Rotate` first argument is radians.
```

```c
#define Matrix_RotateY_s(binang, mtxMode) Matrix_RotateY(BINANG_TO_RAD(binang), mtxMode)
#define Matrix_RotateX_s(binang, mtxMode) Matrix_RotateX(BINANG_TO_RAD(binang), mtxMode)
#define Matrix_RotateZ_s(binang, mtxMode) Matrix_RotateZ(BINANG_TO_RAD(binang), mtxMode)

#define Matrix_RotateY_f(degf, mtxMode)   Matrix_RotateY(DEGF_TO_RADF(degf), mtxMode)
#define Matrix_RotateX_f(degf, mtxMode)   Matrix_RotateX(DEGF_TO_RADF(degf), mtxMode)
#define Matrix_RotateZ_f(degf, mtxMode)   Matrix_RotateZ(DEGF_TO_RADF(degf), mtxMode)
```

`DEGF` \
**f32** rotation value. - 90° would be equal to `90.0f`.

`BINANG` \
**s16** rotation value. - 90° would be equal to `0x4000`.

`RADF` \
**f32** rotation value. - 90° would be equal to `M_PI / 2.0f`.

[Back to Table of Content](#table-of-content)

# Physics
Physics takes care of doing calculations.

#### TODO:
- [x] ``Strand`` Uses one head, used for **hair**, **tails**, etc...
- [ ] ``Bridge`` Same as strand but instead would have two heads for head and tail.
- [ ] ``Cloth`` **Capes**, **dresses**, etc...

## PhysicsStrand Struct

> These are the main struct for drawing dynamic strand. Think of as ``settings``.

Reviewer: ```suggestion
These are the main structs for drawing dynamic strand. Think of them as ``settings``.
```

> Difference between these is that `PhysicsStrandInit` has `flexible array` at the bottom of it, which is perfect for initialization, if you want to keep your initial values stored.

Reviewer: ```suggestion
Difference between these is that `PhysicsStrandInit` has a `flexible array` at the end of it, which is perfect for initialization, if you want to keep your initial values stored.
```

	f32 limbsLength[]; // Flexible array
} PhysicsStrandInit;
```

[Back to Table of Content](#table-of-content)

### PhysicsInfo
Here we have the "settings" for how the physics behave.

```c
typedef struct {
	s32 numLimbs;
	f32 gravity;
	f32 floorY;  // world.pos.y, won't go through this
	f32 maxVel;  // Clamps velocity value
	f32 velStep; // Values below 1.0f will give it spring like motion
	f32 velMult; // Control the power of velocity
} PhysicsInfo;
```

``numLimbs`` \
is the amount of limbs you have for your strand. Basically how many bones you've set to it in Blender.

``gravity`` \
 as you'd think, affects how strongly it'll be pulled upwards or downwards.

``floorY`` \
is the world position, which will prevent the limbs going below that. Bot fear not, if the target goes below this, the limbs wont be stuck at this pos, but they just aim towards this ``Y`` location. This effect is utilized by ``Volvagia`` in Ocarina of Time. When Volvagia goes back into the hole, its mane has natural movement into the hole as well.

``maxVel`` \
clamps the velocity to certain value. If you're not sure what value to use here, safest value would be `50.0f`.

``velStep`` \

> could be also called ___deacceleration step___. It handles from stepping the velocity towards ``0.0f``. But it also works as a __dampening__ value. __velStep__ below ``1.0f`` will have spring like motion and above ``1.0f`` dampens it.

Reviewer: I'm sorry to say but `deacceleration` isn't a word, did you mean acceleration or deceleration 😅

Author: I have difficulties even with my own native language... 😅

Author: I meant deceleration here

``velMult`` \
multiplies the calculation that gets the velocity. It is normally used to soften the velocity, as higher than ``1.0f`` _might_ become unstable. But going above ``1.0f`` isn't bad. Best way to find the results you want is to try.

[Back to PhysicsStrand struct](#physicsstrand-struct)\
[Back to Table of Content](#table-of-content)

### PhysicsHead
Values that are used in the **Physics_DrawDynamicStrand** to have the strand aligned to the head. These can be altered between **Physics_GetHeadProperties** and 

> **Physics_DrawDynamicStrand**, and in some cases it's even necessary. Trial and error is definitely required if working with ``folder characters``.

Reviewer: "folder" characters? you mean folded?

Author: I have no words for this one 😂

```c
typedef struct {
	Vec3f pos;
	Vec3s rot;
	MtxF  mtxF;
} PhysicsHead;
```

> If you'd like to alter position, use Physics_GetHeadProperties ``headPosModel`` to alter position in heads model space. The local axises depend on the `Head direction` in rest pose, the axises might not match being `X = X, Y = Y, Z = Z`.

Reviewer: ```suggestion
If you'd like to alter position, use Physics_GetHeadProperties ``headPosModel`` to alter position in heads model space. The local axises depend on the `Head direction` in rest pose, the axes might not match being `X = X, Y = Y, Z = Z`.
```
